Add TcpProxy::CombinedUpstream for using Router::UpstreamRequest in tcp tunneling path

vikaschoudhary16 opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: This PR is carved out of https://github.com/envoyproxy/envoy/pull/27183#discussion_r1218113408
Additional Description: This PR adds a new upstream in tcpproxy, CombinedUpstream which own a Router::UpstreamRequest instance. There are only refactoring changes to add this new upstream but there are no functional changes as this new upstream is not initialized yet.

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

vikaschoudhary16 commented:
/retest

RyanTheOptimist commented on source/common/tcp_proxy/upstream.h:
```diff
@@ -68,16 +85,243 @@ class HttpConnPool : public GenericConnPool, public Http::ConnectionPool::Callba
+  void onHttpPoolReady(Upstream::HostDescriptionConstSharedPtr& host,
+                       Ssl::ConnectionInfoConstSharedPtr ssl_info);
+
+  struct NullHedgePolicy : public Router::HedgePolicy {
```
Nit here and elsewhere please add a brief comment which explains the purpose of this object. Also Router::HedgePolicy is a class, so the subclass should be a class as well (and hence have private member)

RyanTheOptimist commented on source/common/tcp_proxy/upstream.h:
```diff
@@ -221,21 +528,43 @@ class HttpUpstream : public GenericUpstream, protected Http::StreamCallbacks {
+  void setRouterUpstreamRequest(UpstreamRequestPtr) override {}
+};
+
+class CombinedUpstream : public HttpUpstream {
```
Please add a descriptive comment which explains what this class is all about.

vikaschoudhary16 replied:
done

RyanTheOptimist commented on source/common/tcp_proxy/upstream.h:
```diff
@@ -68,16 +85,243 @@ class HttpConnPool : public GenericConnPool, public Http::ConnectionPool::Callba
+    bool empty() const override { return true; }
+
+    static const std::vector<std::reference_wrapper<const Router::RateLimitPolicyEntry>>
+        rate_limit_policy_entry_;
```
What's the story with this static member? I'm confused by all of these Null structs and global variables. This seems  a bit scary to me.

vikaschoudhary16 replied:
UpstreamRequest accepts `RouterFilterInterface`, so all these Null structs and global variables are to implement RouterFilterInterface which is passed to initialize Router::UpstreamRequest. Many of these will not be used in the tunneling flow. 
I looked at how async client is implemeting RouterFilterInterface, https://github.com/envoyproxy/envoy/blob/main/source/common/http/async_client_impl.h#L156 and following exactly same

RyanTheOptimist replied:
Instead of copying all of the NullFoo classes from async_client_impl.h, what would you think of moving those classes somewhere they can be shared?

vikaschoudhary16 replied:
done.

RyanTheOptimist reviewed (commented):
As a high level comment, I'm a bit unclear how this PR relates to the discussion here https://github.com/envoyproxy/envoy/pull/27183#discussion_r1218141794 
I thought that discussion was suggesting that the HTTP/1 codec be changes but that doesn't seem to be happening here?

vikaschoudhary16 commented:
> As a high level comment, I'm a bit unclear how this PR relates to the discussion here [#27183 (comment)](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1218141794) I thought that discussion was suggesting that the HTTP/1 codec be changes but that doesn't seem to be happening here?

Hi @RyanTheOptimist, Thanks a lot for taking a look. 

Apologies that I linked incorrect comment. Here is the[ comment](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1229922240) that I wanted to link

RyanTheOptimist commented:
> > As a high level comment, I'm a bit unclear how this PR relates to the discussion here [#27183 (comment)](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1218141794) I thought that discussion was suggesting that the HTTP/1 codec be changes but that doesn't seem to be happening here?
> 
> Hi @RyanTheOptimist, Thanks a lot for taking a look.
> 
> Apologies that I linked incorrect comment. Here is the[ comment](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1229922240) that I wanted to link

Ah. Thanks for the link. That makes more sense. I look forward to seeing this PR with additional comments. It's rather opaque at the moment. In particular the PR description says that CombinedUpstream own a Router::UpstreamRequest instance but the class seems to have no members, and "Combined" is a curious name which, from the comment, seems to refer to "combining the H1 and H2 upstreams" but I don't quite understand what that has to do with owning a Router::UpstreamRequest. Hopefully the comments will make this clearer.

RyanTheOptimist commented:
/wait

vikaschoudhary16 commented:
> > > As a high level comment, I'm a bit unclear how this PR relates to the discussion here [#27183 (comment)](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1218141794) I thought that discussion was suggesting that the HTTP/1 codec be changes but that doesn't seem to be happening here?
> > 
> > 
> > Hi @RyanTheOptimist, Thanks a lot for taking a look.
> > Apologies that I linked incorrect comment. Here is the[ comment](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1229922240) that I wanted to link
> 
> Ah. Thanks for the link. That makes more sense. I look forward to seeing this PR with additional comments. It's rather opaque at the moment. In particular the PR description says that CombinedUpstream own a Router::UpstreamRequest instance but the class seems to have no members, and "Combined" is a curious name which, from the comment, seems to refer to "combining the H1 and H2 upstreams" but I don't quite understand what that has to do with owning a Router::UpstreamRequest. Hopefully the comments will make this clearer.

Updated more detailed description and comments. PTAL if still need more details.

vikaschoudhary16 commented:
I am working to get coverage passing. Other tests are passing

phlax commented:
/retest

RyanTheOptimist commented:
> > > > As a high level comment, I'm a bit unclear how this PR relates to the discussion here [#27183 (comment)](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1218141794) I thought that discussion was suggesting that the HTTP/1 codec be changes but that doesn't seem to be happening here?
> > > 
> > > 
> > > Hi @RyanTheOptimist, Thanks a lot for taking a look.
> > > Apologies that I linked incorrect comment. Here is the[ comment](https://github.com/envoyproxy/envoy/pull/27183#discussion_r1229922240) that I wanted to link
> > 
> > 
> > Ah. Thanks for the link. That makes more sense. I look forward to seeing this PR with additional comments. It's rather opaque at the moment. In particular the PR description says that CombinedUpstream own a Router::UpstreamRequest instance but the class seems to have no members, and "Combined" is a curious name which, from the comment, seems to refer to "combining the H1 and H2 upstreams" but I don't quite understand what that has to do with owning a Router::UpstreamRequest. Hopefully the comments will make this clearer.
> 
> Updated more detailed description and comments. PTAL if still need more details.

AH! This is so much more clear. Thank you very much for doing this. Ok, let me dive into the code.

RyanTheOptimist commented on source/common/http/async_client_impl.cc:
```diff
@@ -16,22 +16,20 @@ namespace Envoy {
+const std::multimap<std::string, std::string> RouteEntryImpl::opaque_config_;
+const NullPathMatchCriterion RouteEntryImpl::path_match_criterion_;
+const RouteEntryImpl::ConnectConfigOptRef RouteEntryImpl::connect_config_nullopt_;
+const std::list<LowerCaseString> NullCommonConfig::internal_only_headers_;
```
nit: I believe these are all the static global variables associated with the classes in null_route_impl.h. Let's create null_route_impl.cc and move them there.

vikaschoudhary16 replied:
done

RyanTheOptimist commented on source/common/http/null_route_impl.h:
```diff
@@ -0,0 +1,242 @@
+namespace Envoy {
+namespace Http {
+
+struct NullHedgePolicy : public Router::HedgePolicy {
```
nit: Let's add a comment at the top of this file explaining that this class is all about providing a no-op implementation of Router::Route which requires no-op implementations of a variety or Router helper classes.

vikaschoudhary16 replied:
done

RyanTheOptimist commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -462,6 +475,96 @@ class Filter : public Network::ReadFilter,
   };
 
   StreamInfo::StreamInfo& getStreamInfo();
+  class HttpStreamDecoderFilterCallbacks : public Http::StreamDecoderFilterCallbacks,
```
Class comment, please.

vikaschoudhary16 replied:
done

RyanTheOptimist reviewed (commented):
Ok, I'm starting to get my head around this PR. But I'm unclear about something. The description talks about implementing CombinedUpstream but not initializing it. I'm not sure I see where the new CombinedUpstream is used/created. The only occurrences of the string "CombinedUpstream" in source/ seem to be in source/common/tcp_proxy/upstream.h and source/common/tcp_proxy/upstream.cc where the class is declared and defined. Where is it used?

Relatedly, there seems to be a lot of refactoring that I don't quite understand the motivation for. For example, the StreamDecoderFilterCallbacks refactor. Or why HttpUpstream now needs to implement Router::RouterFilterInterface.

I think this is moving in the right direction but I'm not fully on top of it yet.

vikaschoudhary16 commented on source/common/tcp_proxy/upstream.h:
```diff
@@ -68,16 +86,22 @@ class HttpConnPool : public GenericConnPool, public Http::ConnectionPool::Callba
       ASSERT(conn_pool_ != nullptr);
-      conn_pool_->onGenericPoolReady(host_, request_encoder.getStream().connectionInfoProvider(),
-                                     ssl_info_);
+      if (request_encoder != nullptr) {
```
request_encoder will be null in case CombinedUpstream is handling the stream.  tcp_proxy will not be managing request_encoder in case of CombinedStream/Router::UpstreamRequest.  In this PR, rest of the changes I skipped because that would have required feature gate. Please refer[ here](https://github.com/envoyproxy/envoy/pull/27183/files#diff-8b2f11a21a008c86b7a6203cd94bd8b94455fe9ca4e9c04a52537c0233b3b9afR316) in the parent PR for complete changes

vikaschoudhary16 commented:
> implementing CombinedUpstream but not initializing it. I'm not sure I see where the new CombinedUpstream is used/created. The only occurrences of the string "CombinedUpstream" in source/ seem to be in source/common/tcp_proxy/upstream.h and source/common/tcp_proxy/upstream.cc where the class is declared and defined. Where is it used?

Intentionally it is not being used in this PR because thats what Alyssa suggested for. In the parent PR, you can see CombinedStream getting initialized in [HttpConnPool::newStream()](https://github.com/envoyproxy/envoy/pull/27183/files#diff-3ef49156d77fb763f2ec4f96dc0b8e0048ba095314cc8a972f344848f38777d7R322) and there itself CombinedUpstream is being used just after initialization( `upstream_->newStream(callbacks)`)

vikaschoudhary16 commented:
>For example, the StreamDecoderFilterCallbacks refactor. 

StreamDecoderFilterCallbacks are the callbacks which UpstreamRequest uses to get downstream connection info for setting up upstream and also to communicate back for flow control etc. I have updated class comment as well. Please let me know if still not clear.

>Or why HttpUpstream now needs to implement Router::RouterFilterInterface.

UpstreamRequest's constructor needs RouterFilterInterface. In HCM case, it is the `Router` filter who initializes/owns the UpstreamRequest and pass itself as the RouterFilterInterface to the UpstreamRequest constructor. In tcp_proxy case, instead of Router filter, `TcpProxy::HttpUpstream` is doing the same role of managing UpstreamRequest, hence implemented RouterFilterInterface through HttpUpstream

vikaschoudhary16 commented:
/retest

vikaschoudhary16 commented:
/retest

phlax commented on ci/do_ci.sh:
```diff
@@ -4,7 +4,6 @@
 
 set -e
 
-
```
these changes seem unrelated to this pr - please revert

vikaschoudhary16 replied:
Accidentally this file got pushed while merging main branch. Only realized after you pointed. Thanks a lot!

phlax reviewed (changes requested):
hi @vikaschoudhary16 i dont think the reformatting here is an improvemnt, but regardless i dont think they should be part of this PR

vikaschoudhary16 commented:
friendly ping @RyanTheOptimist :)

RyanTheOptimist reviewed (approved):
Sorry for the delay! The C++ all looks good here, but I'd like to make sure this gets a review from someone with a bit more first-hand experience of the affected classes, so I'll hand off to @mattklein123

RyanTheOptimist reviewed (commented):
/assign @mattklein123

RyanTheOptimist reviewed (changes requested):
Marking "request changes" to make it clear that this is waiting for Matt's approval :)

ohadvano commented:
/assign @mattklein123

mattklein123 commented:
I'm going to reassign to @alyssawilk who I think is sporadically around as she is most familiar with this code and I think it would be good to get a review from her.

alyssawilk commented:
Hey @vikaschoudhary16 , I am back but around interittantly.  I'll try to do apass tomorrow but I thought the first split PR was just going to be combining the HTTP1 and HTTP2 upstreams.  Looks like this does a lot more.  Can we do the simple combine, and do any other refactors in a follow-up?

vikaschoudhary16 commented:
> Hey @vikaschoudhary16 , I am back but around interittantly. I'll try to do apass tomorrow but I thought the first split PR was just going to be combining the HTTP1 and HTTP2 upstreams. Looks like this does a lot more. Can we do the simple combine, and do any other refactors in a follow-up?

Hi @alyssawilk , Thanks for taking a look!
Removed most of the refactoring other than  CombinedUpstream. Only refactoring left is extracting out NullRouteImpl from async client. I thought it is pretty isolated and not a complex refactoring. Please let me know if this also needs to be excluded.

vikaschoudhary16 commented:
/retest

alyssawilk commented:
I'd definitely leave that split out as well - at least I don't think it's functionally necessary for this PR and I'm not (yet) clear why we need it at all.  If it is necessary I'd be inclined to split it out as well though
/wait

vikaschoudhary16 commented:
> I'd definitely leave that split out as well - at least I don't think it's functionally necessary for this PR and I'm not (yet) clear why we need it at all. If it is necessary I'd be inclined to split it out as well though /wait

Alright, reverted this async client refactoring. PTAL now :)

vikaschoudhary16 commented:
had to force push last commit to fix DCO.

vikaschoudhary16 commented:
@alyssawilk sorry for the delay in updating the PR.

I have cleaned up the PR. Now it has bits required by CombinedStream only.
coverage is failing because changes are added to make HttpUpstream implement RouterFilterInterface. 
 In the previous commits, I refactored NullRouteImpl out of async client and reused in CombineUpstream which made coverage improve. But you asked to remove that refactoring, so coverage is failing now. 
If you are strict on not adding NullRouteImpl refactoring in this PR, I can open a separate PR to do that refacting and then rebase this one to use that to improve coverage. Please guide here

vikaschoudhary16 commented on ci/do_ci.sh:
```diff
@@ -721,6 +721,7 @@ case $CI_TARGET in
              --config=rbe-toolchain-tsan \
              -c dbg \
              --build_tests_only \
+             --test_timeout=390,750,1500,5700 \
```
done

alyssawilk commented on source/common/tcp_proxy/upstream.cc:
```diff
@@ -11,6 +12,24 @@
 namespace Envoy {
 namespace TcpProxy {
+
+const std::vector<std::reference_wrapper<const Router::RateLimitPolicyEntry>>
```
sorry to keep coming back to this but if all we're doing is combining the http1 and http2 upstreams into one class I don't think we should need all of this
/wait

vikaschoudhary16 replied:
This new class owns a `Router::UpstreamRequest` instance. Unlike existing `Http1Upstream` and `Http2Upstream`, new class handles traffic generically by invoking appropriate functions on `Router::UpstreamRequest`. To test these methods, tests need to initialize `Router::UpstreamRequest` and therefore now test are using `NullRouteImpl` lib.

alyssawilk reviewed (commented):
Sorry I'm still seeing changes here I don't think are necessary.  Why do we need to do anything with routes in a pure refactor PR?
I really would love a PR that's just combining the two upstreams, no other chnages.
/wait

ohadvano commented:
Drive by comment as I'm tracking this PR, @alyssawilk when saying combining do you mean having a single class for ``HttpUpstream`` that has all the logic to switch per upstream codec inside rather than having ``Http1Upstream`` + ``Http2Upstream`` + ``CombinedUpstream (for the new case)``?

vikaschoudhary16 commented:
>Sorry I'm still seeing changes here I don't think are necessary. Why do we need to do anything with routes in a pure refactor PR?

@alyssawilk This is not a pure refactor PR. This PR is adding a new upstream called "CombinedUpstream" without changing behavior of existing upstreams, `Http1Upstream` and `Http2Upstream`. New class is not "combining" functionalities of existing  `Http1Upstream` and `Http2Upstream`. Sorry if I chosen a bad name for newly added class.

alyssawilk commented:
If you're changing code without a functional change it's a refactor PR.  I think you can combine the two upstreams without adding route logic.  Please back those changes out or explain why they're necessary?  + @KBaichoo for a first pass review

vikaschoudhary16 commented:
@alyssawilk I added new upstream, `CombinedUpsteam` , as per your guidance in [this comment](https://github.com/envoyproxy/envoy/pull/27183/commits/5207356084c22ad8273ad6eaa88d09b3c764ef6e#r1188715644)

>I think in the new path we can merge the HTTP1 and HTTP2 upstreams, as the upstream codec filter should take care of per codec logic. so maybe have a new HttpUpstream which handles both and it can call newStream when created?

Existing `TcpProxy::Http1Upstream` and `TcpProxy::Http2Upstream` owns respective codecs directly and forwards header/data through invoking encodeXXX() functions directly on their respective request encoders.

In newly introduced `TcpProxy::CombinedStream`, it owns a `Router::UpstreamRequest` instance. Then in combined manner(irrespective of h1 or h2), it invokes methods on  `Router::UpstreamRequest` instance, example [acceptDataFromRouter()](https://github.com/envoyproxy/envoy/pull/27991/files#diff-3ef49156d77fb763f2ec4f96dc0b8e0048ba095314cc8a972f344848f38777d7R357),  to forward traffic to upstream.
![image](https://github.com/envoyproxy/envoy/assets/11936386/531a8932-2770-4661-a100-ca930ba4fb87)


```
void CombinedUpstream::encodeData(Buffer::Instance& data, bool end_stream) {
  if (upstream_requests_.empty()) {
    return;
  }
  upstream_requests_.front()->acceptDataFromRouter(data, end_stream);
  if (end_stream) {
    doneWriting();
  }
}
```

Since CombinedUpstream's method's definition is using `Router::UpstreamRequest`, to unit test these methods we need to initialize an instance of `Router::UpstreamRequest`. Constructor of  `Router::UpstreamRequest` expects a `Router::RouterFilterInterface`. 
Coming back to question that why we are using route logic, RouteFilterInterface declares a method `route()`. So TcpProxy::HttpUpstream need to define `route()` because it implements RouterFilterInterface. `route()` when invoked on HttpUpstream returns `Router::NullRouteImpl` instance.

In the tests to initialize HttpUpstream, Router::NullRouteImpl is being initialized and passed to HttpUpstream.

alyssawilk commented:
Ok, I think we're getting down to the root of the confusion here.

The first Pr I was asking for was literally just removing the Http1Upstream and Http2Upstream and moving that code into the HttpUpstream base class.  I was hoping all other changes (including ownership of the UpstreamRequest) could go into follow-ups.

Refactoring the data plane tends to result in high risk changes, and I'd prefer to make changes smaller and safer where possible.  I'm willing to let @KBaichoo be tie breaker here - if he's happy reviewing the PR as is and comfortable there's functionally no risk of breakage you can leave this as is.  If he prefers the split we'll go with that.

/wait

KBaichoo reviewed (commented):
Thank you for the neat diagram @vikaschoudhary16!

I agree with Alyssa here that we should split out the refactor changes vs the
new functional changes here. Having an integration test would also help so
that we can gain additional confidence. The CombinedUpstream AFAICT isn't
instantiated in any of the non-test code and it'd be helpful to check it in
in a point where it's used.
